### PR TITLE
feat: Working day and holidays calculator (CMC-510)

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link (if applicable) ###
[CMC-510](url)

### Change description ###
This PR adds the services that can be reused to calculate different dates for case based on working days and holidays across the year.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    Division englandAndWales;
+
+    static final class Countries {
+
+        static final String ENGLAND_AND_WALES = "england-and-wales";
+
+        private Countries() {
+            // NO-OP
+        }
+
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java]: can remove blank line

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidaysApi.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@FeignClient(name = "bank-holidays-api", url = "${bankHolidays.api.url}")
+public interface BankHolidaysApi {
+
+    @GetMapping(path = "/bank-holidays.json")
+    BankHolidays retrieveAll();
+}
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidaysApi.java:11]: thoughts on removing this from unspec and placing in it's own calendar / bank holidays folder?

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    static class Division {
+        @JsonProperty("events")
+        List<EventDate> events;
+
+        @JsonIgnoreProperties(ignoreUnknown = true)
+        static class EventDate {
+            static final String FORMAT = "yyyy-MM-dd";
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java]: where is this used?

Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java]: ahh ok it is used later on. why not always return LocalDate? could use JsonFormat annotation?

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollection.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    private Set<LocalDate> retrieveAllPublicHolidays() {
+        BankHolidays value = bankHolidaysApi.retrieveAll();
+
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(BankHolidays.Division.EventDate.FORMAT);
+
+        return value.englandAndWales.events.stream()
+            .map(item -> LocalDate.parse(item.date, formatter))
+            .collect(Collectors.toSet());
+    }
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollection.java]: this becomes simpler if you abstract the formatting, can chain and stream directly on retrieveAll method

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            assertTrue(service.isWorkingDay(weekday));
+        }
+    }
+
+    @Nested
+    class ForWeekend {
+
+        @ParameterizedTest
+        @ArgumentsSource(Weekend.class)
+        void shouldReturnFalseF_whenItsWeekend(LocalDate weekend) {
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]:
```suggestion
        void shouldReturnFalse_whenWeekend(LocalDate weekend) {
```

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    void setup() {
+        service = new WorkingDayIndicator(publicHolidaysApiClient, nonWorkingDaysCollection);
+    }
+
+    @Nested
+    class ForWeekdays {
+
+        @ParameterizedTest
+        @ArgumentsSource(WeekDays.class)
+        void shouldReturnTrue_whenItsWeekdays(LocalDate weekday) {
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]:
```suggestion
        void shouldReturnTrue_whenWeekday(LocalDate weekday) {
```

File: src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/NonWorkingDaysCollectionTest.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+class NonWorkingDaysCollectionTest {
+
+    private NonWorkingDaysCollection collection;
+
+    @Test
+    void shouldReturnTrue_whenMatchingNonWorkingDays() {
+        collection = new NonWorkingDaysCollection("/non-working-days/nwd-valid.dat");
+        assertThat(collection.contains(LocalDate.of(2020, Month.DECEMBER, 2))).isTrue();
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/NonWorkingDaysCollectionTest.java]: can use assertTrue / assertFalse for all assertions in this file

File: src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    void shouldReturnAllBankHolidays_whenSuccessfulResponse() {
+        when(bankHolidaysApi.retrieveAll()).thenReturn(createExpectedResponse());
+        PublicHolidaysCollection publicHolidaysCollection = new PublicHolidaysCollection(bankHolidaysApi);
+
+        Set<LocalDate> response = publicHolidaysCollection.getPublicHolidays();
+
+        assertAll(
+            "Bank holidays",
+            () -> assertThat(response).isNotEmpty(),
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java]: notEmpty assertion adds nothing here, I would remove

File: src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    void shouldMakeSingleExternalApiCall_whenCacheIsNotPopulated() {
+        when(bankHolidaysApi.retrieveAll()).thenReturn(createExpectedResponse());
+        PublicHolidaysCollection publicHolidaysCollection = new PublicHolidaysCollection(bankHolidaysApi);
+
+        Set<LocalDate> resultFromApi = publicHolidaysCollection.getPublicHolidays();
+        Set<LocalDate> resultFromCache = publicHolidaysCollection.getPublicHolidays();
+        Set<LocalDate> resultFromCacheAgain = publicHolidaysCollection.getPublicHolidays();
+
+        verify(bankHolidaysApi, times(1)).retrieveAll();
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java]: times(1) is default, can remove

File: src/test/resources/non-working-days/nwd-invalid.dat [hunk 10]
```diff
@@ -0,0 +1,3 @@
+2019
+I am a teapot
+¯\_(ツ)_/¯
```
Comment (swalker125) [inline src/test/resources/non-working-days/nwd-invalid.dat:3]: 😂

**Review by swalker125 — COMMENTED** [review]
quick flick through, couple of things

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidaysApi.java (hunk 2 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidaysApi.java:11]: Probably moving this to a submodule would be another option. Will do that afterwards as there are branches out of this branch.

File: src/main/java/uk/gov/hmcts/reform/unspec/Application.java [hunk 11]
```diff
@@ -7,7 +7,7 @@
 
 @SpringBootApplication
 @EnableCircuitBreaker
-@EnableFeignClients(basePackages = {"uk.gov.hmcts.reform.idam.client"})
+@EnableFeignClients(basePackages = {"uk.gov.hmcts.reform.idam.client", "uk.gov.hmcts.reform.unspec"})
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/Application.java:10]: is this required?

File: src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        requireNonNull(date);
+
+        return isWorkingDay(date) ? date : getNextWorkingDay(date.plusDays(1));
+    }
+
+    public LocalDate getPreviousWorkingDay(LocalDate date) {
+        requireNonNull(date);
+
+        return isWorkingDay(date) ? date : getPreviousWorkingDay(date.minusDays(1));
+    }
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java:52]: is this needed?

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+            @JsonProperty("date")
+            String date;
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java]: comment resolved but this is still a string, if it was a date then follow on logic would be simpler.

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
+package uk.gov.hmcts.reform.unspec.service.bankholidays;
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java:1]: still think this should exist separate to unspec directory

Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/BankHolidays.java:1]: ok seen you will do in separate PR

File: src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollection.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Autowired
+    public PublicHolidaysCollection(BankHolidaysApi bankHolidaysApi) {
+        this.bankHolidaysApi = bankHolidaysApi;
+    }
+
+    private Set<LocalDate> retrieveAllPublicHolidays() {
+        BankHolidays value = bankHolidaysApi.retrieveAll();
+
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(BankHolidays.Division.EventDate.FORMAT);
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollection.java]: still think this should exist in BankHoliday model

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+    @BeforeEach
+    void setup() {
+        service = new WorkingDayIndicator(publicHolidaysApiClient, nonWorkingDaysCollection);
+    }
+
+    @Nested
+    class ForWeekdays {
+
+        @ParameterizedTest
+        @ArgumentsSource(WeekDays.class)
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]: could actually use enum source annotation here, probably a little less code. can then specify which days using name in combo with include / exclude

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+            assertFalse(service.isWorkingDay(weekend));
+        }
+    }
+
+    @Nested
+    class IsWorkingDay {
+
+        @Test
+        void shouldReturnFalseForOneBankHoliday_whenThereIsOneBankHolidayInCollection() {
+            LocalDate bankHoliday = BANK_HOLIDAY;
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]: why not use BANK_HOLIDAY directly?

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+
+            LocalDate nextWorkingDay = service.getNextWorkingDay(BANK_HOLIDAY);
+
+            assertEquals(NEXT_WORKING_DAY_AFTER_BANK_HOLIDAY, nextWorkingDay);
+        }
+
+    }
+
+    @Nested
+    class PreviousWorkingDay {
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java:170]: not sure of the use case for this

File: src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    void shouldReturnAllBankHolidays_whenSuccessfulResponse() {
+        when(bankHolidaysApi.retrieveAll()).thenReturn(createExpectedResponse());
+        PublicHolidaysCollection publicHolidaysCollection = new PublicHolidaysCollection(bankHolidaysApi);
+
+        Set<LocalDate> response = publicHolidaysCollection.getPublicHolidays();
+
+        assertAll(
+            "Bank holidays",
+            () -> assertThat(response.contains(toDate(BANK_HOLIDAY_1))),
+            () -> assertThat(response.contains(toDate(BANK_HOLIDAY_2)))
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/service/bankholidays/PublicHolidaysCollectionTest.java]: this looks like a false positive, assertThat(response).contains...

File: src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java (hunk 12 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java:52]: May not be in the current stories in had. but we might need in future stories. We needed this in OCMC, so will have similar use cases here as well. @ckalista please advise accordingly.

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java (hunk 17 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]: yes we can but was separating concerns based on [Arrange-Act-Assert](https://github.com/testdouble/contributing-tests/wiki/Arrange-Act-Assert)  steps.

File: src/main/java/uk/gov/hmcts/reform/unspec/Application.java (hunk 11 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/Application.java:10]: yes this is required.

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java (hunk 16 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java]: There is no enum here, passing a list of dates that are weekdays here.

File: src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java (hunk 12 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java:52]: we should as we need in my opinion, no point having dead code sat around

Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicator.java:52]: clarified with Chris and we are going to keep this in.

File: src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java (hunk 18 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/service/WorkingDayIndicatorTest.java:170]: as above response

**Review by swalker125 — APPROVED** [review]

**Comment by swalker125** [comment]
still think BankHolidays model should return LocalDate and not String, but not the end of the world I guess

**Review by dharmendrak — APPROVED** [review]

**Pull request merged**